merlin-sdk/src/main/java/gov/nasa/jpl/aerie/merlin/protocol/json/SerializedValueSerializer.java

```diff
@@ -0,0 +1,110 @@
+package gov.nasa.jpl.aerie.merlin.protocol.json;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.ser.std.StdSerializer;
+import gov.nasa.jpl.aerie.merlin.protocol.types.SerializedValue;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public final class SerializedValueSerializer extends StdSerializer<SerializedValue> {
```

Comment:
The purpose of the `merlin.protocol` package is to coordinate communication between mission models and systems that use them (like the simulation engine). It's not clear, but I don't think this class is used to that end.

I'm not fond of giving all models a transitive dependency on Jackson; is it possible to move these serialization classes out of `merlin.package`?

Reply:
I'm not sure what you mean by transitive dependency.

In Jackson, the Java class you want to serialize or deserialize defines its own schema. If the JSON schema is different from the Java schema, you usually put annotations on the class that tell Jackson what's different. If the schemas are _too_ different for annotations to handle (which is rare and only happened for SerializedValue), you can write a custom serializer and deserializer. But, the (de)serializer is attached to the class using

```java
@JsonSerialize(using = SerializedValueSerializer.class)
@JsonDeserialize(using = SerializedValueDeserializer.class)
public interface SerializedValue {
    ...
```

The JSON schema is attached to the class. When you want to deserialize it, you use `mapper.readValue(inputString, SerializedValue.class)` just like you would do with a class that doesn't have custom a serializer or deserializer.

Reply:
Ok I figured out what you said. Yes it is possible to remove the `merlin.protocol` dependency on Jackson, but in my opinion its a worse solution, and the reason is actually because of my original comment.

The serializer and deserializer are attached to the SerializedValue class, so that every other package attempting to communicate with SerializedValues can just call `mapper.readValue(input, SerializedValue.class)`, without even knowing that it has a custom serializer and deserializer. If you move the (de)serializer to another package, they can't be attached to the SerializedValue class, and you'll have to associate them at runtime when you create the object mapper. Which would usually look like this:

```java
private static final ObjectMapper mapper = new ObjectMapper();
static {
    SimpleModule module = new SimpleModule();
    module.addSerializer(SerializedValue.class, new SerializedValueSerializer());
    module.addDeserializer(SerializedValue.class, new SerializedValueDeserializer());
    mapper.registerModule(module);
}
```

And you'll have to remember to do that in every package that creates an ObjectMapper. That's a lot more prone to error, so I'd prefer not to do it.

Reply:
I understand the alternatives regarding making Jackson aware of these types. But, we would be in the same situation if we had a third-party library whose types we needed to serialize, and whose source we could not edit.

We can provide a `MerlinJacksonModule` class in the `merlin.driver` package like the below, and if downstream clients want to use Jackson to serialize Merlin types, they can call `mapper.registerModule(MerlinJacksonModule.get())`. If you forget to do this, you'll find out whether you got it wrong as soon as you try to deserialize something or use the serialized data.

```java
public final class MerlinJacksonModule {
  public static SimpleModule get() {
    final var module = new SimpleModule();
    module.addSerializer(SerializedValue.class, new SerializedValueSerializer());
    module.addDeserializer(SerializedValue.class, new SerializedValueDeserializer());
    // and similar for Duration, maybe
    return module;
  }
}

// elsewhere:
mapper.registerModule(MerlinJacksonModule.class);
```

> And you'll have to remember to do that in every package that creates an ObjectMapper. That's a lot more prone to error, so I'd prefer not to do it.

I'd expect to provide top-level methods for parsing entities of interest, like constraints, which would reduce the need for multiple clients to manage their own Jackson parsers. Jackson should be an implementation detail localized to the few sites that need to serialize or deserialize data.